app/models/project_step.rb

```diff
@@ -135,6 +138,27 @@ def handle_original_date_logic
     end
   end
 
+  def handle_finalized_at
+    if is_finalized && ! finalized_at
+      self.finalized_at = Time.now
+    elsif ! is_finalized && finalized_at
+      self.finalized_at = nil
+    end
+  end
```

Comment:
Also a test checking that this works as expected on save would be a good idea I think.